NewForce-at-Mountwest/bangazon-api-the-thunder-lobsters
# Lc customer

#Description
Added functionality to GET, POST, and PUT Customers from the Bangazon API. Added a test suite for integration testing of this functionality. Added string queries to the GET methods for this resource.

Related Ticket(s)
https://trello.com/c/79sCrlHp/13-ticket-1-allow-developers-to-access-the-customer-resource

Steps to Test
Pull down these changes on a new branch
Open a new SQL query for your database
Copy the following code into the new SQL query and execute it:

INSERT INTO Customer (FirstName, LastName, CreationDate, LastActiveDate) VALUES ('Dylan','Bisop', '2020-08-16 11:46:23', '2020-08-16 11:46:23');
INSERT INTO Customer (FirstName, LastName, CreationDate, LastActiveDate) VALUES ('Lindsey', 'Crittendon', '2020-08-17 11:46:23', '2020-08-16 11:46:23');
INSERT INTO Customer (FirstName, LastName, CreationDate, LastActiveDate) VALUES ('Derek', 'Mayse', '2020-08-18 11:46:23', '2020-08-16 11:46:23');
INSERT INTO Customer (FirstName, LastName, CreationDate, LastActiveDate) VALUES ('Barry', 'Griffith', '2020-08-19 11:46:23', '2020-08-16 11:46:23');
INSERT INTO Customer (FirstName, LastName, CreationDate, LastActiveDate) VALUES ('Ashon', 'Woodbury', '2020-08-16 11:46:23', '2020-08-16 11:46:23');
INSERT INTO Customer (FirstName, LastName, CreationDate, LastActiveDate) VALUES ('Potato', 'Derek', '2020-08-15 11:46:23', '2020-08-16 11:46:23');
INSERT INTO Customer (FirstName, LastName, CreationDate, LastActiveDate) VALUES ('Michelle', 'Derek', '2020-08-15 11:46:23', '2020-08-16 11:46:23');

INSERT INTO Product (ProductTypeId, CustomerId, Price, Title, Description, Quantity) VALUES (1, 1, 1, 'Test One', 'Test One Description', 1);
INSERT INTO Product (ProductTypeId, CustomerId, Price, Title, Description, Quantity) VALUES (2, 2, 2, 'Test Two', 'Test Two Description', 2);

INSERT INTO ProductType (Name) VALUES ('Type One');
INSERT INTO ProductType (Name) VALUES ('Type Two');

INSERT INTO PaymentType (AcctNumber, Name, CustomerId) VALUES ('slgslj', 'Visa', 1);
INSERT INTO PaymentType (AcctNumber, Name, CustomerId) VALUES ('bloop', 'Mastercard', 2);

If you don't have an appsettings.json file, create one in the BangazonAPI project and paste the following code:
{ "Logging": { "LogLevel": { "Default": "Warning" } }, "AllowedHosts": "*", "ConnectionStrings": { "DefaultConnection": "Server=localhost\\SQLEXPRESS;Database=BangazonAPI;Trusted_Connection=True;" } }

Edit DatabaseFixture.cs (in TestBangazonAPI) to reference the SQL server running on your machine, rather than SQLEXPRESS.

Press CTRL + F5 to start the API

Change the URL of your browser to end in api/customer if it is not updated automatically

Copy and paste this URL into Postman

Using Postman, ensure that you can GET a list of all customers from this URL. 

Ensure that you can add an ID to the end of the URL to GET a single product resource

Using the default GET method, include '?_include=products', '?_include=payments', and '?q=mic'  string queries.  "mic" can be replaced with any string.

Two customers should have payments and products with those search queries while the other customers do not.

Copy some JSON from a GET result and use it to POST a new customer

Use PUT to edit the product you just POSTed

Use GET to ensure this customer was edited

In Visual Studio, navigate to the Test Explorer

Check the models in DatabaseFixture.cs for TestCustomer against your database. Make sure the CustomerId and PaymentTypeId reference the proper entries (PaymentTypeId specifically should point to a PaymentType with a name of "Test Name"). Also check the editedCustomer in CustomerTest.cs for the same ids.

Run all tests. Ensure that all 5 CustomerTest tests succeed.

#Checklist:
[] I have commented my code, particularly in hard-to-understand areas
[] I have made corresponding changes to the documentation
[] My changes generate no new errors
[] I have added tests that prove my fix is effective or that my feature works
[] New and existing unit tests pass locally with my changes
